File: src/paru_gui/preferences_manager.py
import os
import logging
import json
from typing import List, Optional, Any, Dict, Union
from enum import Enum
from gi.repository import Gio, Gtk, GObject, Adw, GLib

logging.basicConfig(
    level=logging.INFO,
    format='%(asctime)s - %(levelname)s - %(message)s'
)
logger = logging.getLogger("preferences_manager")

class EditorChoice(Enum):
    SYSTEM_DEFAULT = "system_default"
    GEDIT = "gedit"
    NANO = "nano"
    VIM = "vim"
    EMACS = "emacs"
    VSCODE = "code"
    ATOM = "atom"
    SUBLIME = "subl"
    CUSTOM = "custom"

class UpstreamFrequency(Enum):
    NEVER = "never"
    DAILY = "daily"
    WEEKLY = "weekly"
    MONTHLY = "monthly"
    ON_STARTUP = "on_startup"
    MANUAL = "manual"

class AURConfidenceLevel(Enum):
    PARANOID = "paranoid"
    CONSERVATIVE = "conservative"
    BALANCED = "balanced"
    TRUSTING = "trusting"
    PERMISSIVE = "permissive"

class ThemeMode(Enum):
    SYSTEM = "system"
    LIGHT = "light"
    DARK = "dark"

class UpdateNotificationLevel(Enum):
    NONE = "none"
    CRITICAL = "critical"
    IMPORTANT = "important"
    ALL = "all"

class LogLevel(Enum):
    DEBUG = "debug"
    INFO = "info"
    WARNING = "warning"
    ERROR = "error"

class PreferencesManager(GObject.Object):
    SCHEMA_ID = 'org.gnome.paru-gui'

    __gsignals__ = {
        'preference-changed': (GObject.SignalFlags.RUN_LAST, None, (str, object)),
        'preferences-reset': (GObject.SignalFlags.RUN_LAST, None, ()),
        'preferences-exported': (GObject.SignalFlags.RUN_LAST, None, (str,)),
        'preferences-imported': (GObject.SignalFlags.RUN_LAST, None, ()),
    }

    def __init__(self):
        super().__init__()
        try:
            self.settings = Gio.Settings.new(self.SCHEMA_ID)
            logger.info(f"PreferencesManager initialized with GSettings schema: {self.SCHEMA_ID}")
        except GLib.Error as e:
            logger.critical(f"Failed to initialize GSettings for schema '{self.SCHEMA_ID}': {e}")
            logger.critical("Preferences will not be saved or loaded properly. Ensure schema is installed.")
            self.settings = None

    def _get_setting(self, key: str, default_value: Any = None) -> Any:
        if not self.settings:
            return default_value

        try:
            schema_type = self.settings.get_value(key).get_type().get_type_string()
            if schema_type == 's':
                return self.settings.get_string(key)
            elif schema_type == 'b':
                return self.settings.get_boolean(key)
            elif schema_type == 'i':
                return self.settings.get_int(key)
            elif schema_type == 'as':
                return self.settings.get_strv(key)
            else:
                logger.warning(f"Unknown GSettings type '{schema_type}' for key '{key}'. Returning default.")
                return default_value
        except (GLib.Error, AttributeError) as e:
            logger.error(f"Error getting GSettings key '{key}': {e}. Returning default: {default_value}")
            return default_value

    def _set_setting(self, key: str, value: Any) -> bool:
        if not self.settings:
            return False

        try:
            if isinstance(value, bool):
                self.settings.set_boolean(key, value)
            elif isinstance(value, int):
                self.settings.set_int(key, value)
            elif isinstance(value, str):
                self.settings.set_string(key, value)
            elif isinstance(value, list):
                self.settings.set_strv(key, value)
            else:
                logger.error(f"Unsupported value type for key '{key}': {type(value)}")
                return False

            self.emit('preference-changed', key, value)
            return True
        except GLib.Error as e:
            logger.error(f"Error setting GSettings key '{key}' to '{value}': {e}")
            return False

    def get_preference(self, key: str, default_value: Any = None) -> Any:
        return self._get_setting(key, default_value)

    def set_preference(self, key: str, value: Any) -> bool:
        return self._set_setting(key, value)

    def get_simplified_mode(self) -> bool:
        return self.simplified_mode

    def set_simplified_mode(self, value: bool) -> bool:
        self.simplified_mode = value
        return True

    @property
    def first_run(self) -> bool:
        return self._get_setting('first-run', True)

    @first_run.setter
    def first_run(self, value: bool):
        self._set_setting('first-run', value)

    @property
    def simplified_mode(self) -> bool:
        return self._get_setting('simplified-mode', True)

    @simplified_mode.setter
    def simplified_mode(self, value: bool):
        self._set_setting('simplified-mode', value)

    @property
    def default_editor(self) -> str:
        return self._get_setting('default-editor', 'gedit')

    @default_editor.setter
    def default_editor(self, value: str):
        self._set_setting('default-editor', value)

    @property
    def window_width(self) -> int:
        return self._get_setting('window-width', 1200)

    @window_width.setter
    def window_width(self, value: int):
        self._set_setting('window-width', value)

    @property
    def window_height(self) -> int:
        return self._get_setting('window-height', 800)

    @window_height.setter
    def window_height(self, value: int):
        self._set_setting('window-height', value)

    @property
    def recent_directories(self) -> List[str]:
        return self._get_setting('recent-directories', [])

    @recent_directories.setter
    def recent_directories(self, value: List[str]):
        self._set_setting('recent-directories', value)

    @property
    def max_recent_directories(self) -> int:
        return self._get_setting('max-recent-directories', 10)

    @max_recent_directories.setter
    def max_recent_directories(self, value: int):
        self._set_setting('max-recent-directories', value)

    @property
    def upstream_check_frequency(self) -> int:
        return self._get_setting('upstream-check-frequency', 24)

    @upstream_check_frequency.setter
    def upstream_check_frequency(self, value: int):
        self._set_setting('upstream-check-frequency', value)

    @property
    def upstream_ignore_prerelease(self) -> bool:
        return self._get_setting('upstream-ignore-prerelease', True)

    @upstream_ignore_prerelease.setter
    def upstream_ignore_prerelease(self, value: bool):
        self._set_setting('upstream-ignore-prerelease', value)

    @property
    def upstream_notify_security_only(self) -> bool:
        return self._get_setting('upstream-notify-security-only', False)

    @upstream_notify_security_only.setter
    def upstream_notify_security_only(self, value: bool):
        self._set_setting('upstream-notify-security-only', value)

    @property
    def upstream_priority_platforms(self) -> List[str]:
        return self._get_setting('upstream-priority-platforms', ['github', 'gitlab'])

    @upstream_priority_platforms.setter
    def upstream_priority_platforms(self, value: List[str]):
        self._set_setting('upstream-priority-platforms', value)

    @property
    def aur_show_trust_icons(self) -> bool:
        return self._get_setting('aur-show-trust-icons', True)

    @aur_show_trust_icons.setter
    def aur_show_trust_icons(self, value: bool):
        self._set_setting('aur-show-trust-icons', value)

    @property
    def aur_block_no_votes(self) -> bool:
        return self._get_setting('aur-block-no-votes', False)

    @aur_block_no_votes.setter
    def aur_block_no_votes(self, value: bool):
        self._set_setting('aur-block-no-votes', value)

    @property
    def aur_consider_last_update(self) -> bool:
        return self._get_setting('aur-consider-last-update', True)

    @aur_consider_last_update.setter
    def aur_consider_last_update(self, value: bool):
        self._set_setting('aur-consider-last-update', value)

    @property
    def aur_check_negative_comments(self) -> bool:
        return self._get_setting('aur-check-negative-comments', True)

    @aur_check_negative_comments.setter
    def aur_check_negative_comments(self, value: bool):
        self._set_setting('aur-check-negative-comments', value)

    @property
    def aur_confidence_level(self) -> str:
        return self._get_setting('aur-confidence-level', 'balanced')

    @aur_confidence_level.setter
    def aur_confidence_level(self, value: str):
        self._set_setting('aur-confidence-level', value)

    @property
    def developer_mode(self) -> bool:
        return self._get_setting('developer-mode', False)

    @developer_mode.setter
    def developer_mode(self, value: bool):
        self._set_setting('developer-mode', value)

    @property
    def check_devel_updates(self) -> bool:
        return self._get_setting('check-devel-updates', False)

    @check_devel_updates.setter
    def check_devel_updates(self, value: bool):
        self._set_setting('check-devel-updates', value)

    @property
    def clean_after_build(self) -> bool:
        return self._get_setting('clean-after-build', True)

    @clean_after_build.setter
    def clean_after_build(self, value: bool):
        self._set_setting('clean-after-build', value)

    @property
    def install_debug_packages(self) -> bool:
        return self._get_setting('install-debug-packages', False)

    @install_debug_packages.setter
    def install_debug_packages(self, value: bool):
        self._set_setting('install-debug-packages', value)

    @property
    def show_detailed_warnings(self) -> bool:
        return self._get_setting('show-detailed-warnings', False)

    @show_detailed_warnings.setter
    def show_detailed_warnings(self, value: bool):
        self._set_setting('show-detailed-warnings', value)

    @property
    def show_terminal_panel(self) -> bool:
        return self._get_setting('show-terminal-panel', False)

    @show_terminal_panel.setter
    def show_terminal_panel(self, value: bool):
        self._set_setting('show-terminal-panel', value)

    @property
    def enable_sandboxing(self) -> bool:
        return self._get_setting('enable-sandboxing', True)

    @enable_sandboxing.setter
    def enable_sandboxing(self, value: bool):
        self._set_setting('enable-sandboxing', value)

    @property
    def auto_pgp_fetch(self) -> bool:
        return self._get_setting('auto-pgp-fetch', True)

    @auto_pgp_fetch.setter
    def auto_pgp_fetch(self, value: bool):
        self._set_setting('auto-pgp-fetch', value)

    @property
    def theme_mode(self) -> str:
        return self._get_setting('theme-mode', 'system')

    @theme_mode.setter
    def theme_mode(self, value: str):
        self._set_setting('theme-mode', value)

    @property
    def update_notification_level(self) -> str:
        return self._get_setting('update-notification-level', 'important')

    @update_notification_level.setter
    def update_notification_level(self, value: str):
        self._set_setting('update-notification-level', value)

    @property
    def log_level(self) -> str:
        return self._get_setting('log-level', 'info')

    @log_level.setter
    def log_level(self, value: str):
        self._set_setting('log-level', value)

    @property
    def enable_history_tracking(self) -> bool:
        return self._get_setting('enable-history-tracking', True)

    @enable_history_tracking.setter
    def enable_history_tracking(self, value: bool):
        self._set_setting('enable-history-tracking', value)

    @property
    def history_retention_days(self) -> int:
        return self._get_setting('history-retention-days', 30)

    @history_retention_days.setter
    def history_retention_days(self, value: int):
        self._set_setting('history-retention-days', value)

    def add_recent_directory(self, directory: str):
        recent = self.recent_directories
        if directory in recent:
            recent.remove(directory)
        recent.insert(0, directory)
        recent = recent[:self.max_recent_directories]
        self.recent_directories = recent

    def clear_recent_directories(self):
        self.recent_directories = []

    def get_editor_command(self) -> str:
        editor = self.default_editor
        if editor == "system_default":
            return os.environ.get('EDITOR', 'gedit')
        elif editor == "custom":
            return self._get_setting('custom-editor-command', 'gedit')
        else:
            return editor

    def set_custom_editor(self, command: str):
        self._set_setting('custom-editor-command', command)
        self.default_editor = "custom"

    def reset_to_defaults(self):
        if not self.settings:
            return False

        try:
            for key in self.settings.list_keys():
                self.settings.reset(key)
            self.emit('preferences-reset')
            logger.info("All preferences reset to defaults")
            return True
        except Exception as e:
            logger.error(f"Failed to reset preferences: {e}")
            return False

    def export_preferences(self, file_path: str) -> bool:
        if not self.settings:
            return False

        try:
            preferences = {}
            for key in self.settings.list_keys():
                preferences[key] = self._get_setting(key)

            with open(file_path, 'w') as f:
                json.dump(preferences, f, indent=2)

            self.emit('preferences-exported', file_path)
            logger.info(f"Preferences exported to: {file_path}")
            return True
        except Exception as e:
            logger.error(f"Failed to export preferences: {e}")
            return False

    def import_preferences(self, file_path: str) -> bool:
        if not self.settings:
            return False

        try:
            with open(file_path, 'r') as f:
                preferences = json.load(f)
            
            for key, value in preferences.items():
                if key in self.settings.list_keys():
                    self._set_setting(key, value)
            
            self.emit('preferences-imported')
            logger.info(f"Preferences imported from: {file_path}")
            return True
        except Exception as e:
            logger.error(f"Failed to import preferences: {e}")
            return False

    def validate_preferences(self) -> Dict[str, List[str]]:
        errors = {}

        if self.window_width < 800:
            errors.setdefault('window', []).append("Window width must be at least 800 pixels")

        if self.window_height < 600:
            errors.setdefault('window', []).append("Window height must be at least 600 pixels")

        if self.upstream_check_frequency < 0:
            errors.setdefault('upstream', []).append("Check frequency cannot be negative")

        if self.max_recent_directories < 1:
            errors.setdefault('recent', []).append("Must keep at least 1 recent directory")

        if self.history_retention_days < 1:
            errors.setdefault('history', []).append("History retention must be at least 1 day")

        return errors

    def apply_theme(self):
        theme = self.theme_mode
        if theme == "system":
            return

        style_manager = Adw.StyleManager.get_default() if hasattr(Adw, 'StyleManager') else None
        if style_manager:
            if theme == "dark":
                style_manager.set_color_scheme(Adw.ColorScheme.FORCE_DARK)
            elif theme == "light":
                style_manager.set_color_scheme(Adw.ColorScheme.FORCE_LIGHT)
            else:
                style_manager.set_color_scheme(Adw.ColorScheme.DEFAULT)

    def get_all_preferences(self) -> Dict[str, Any]:
        if not self.settings:
            return {}

        preferences = {}
        for key in self.settings.list_keys():
            preferences[key.replace('-', '_')] = self._get_setting(key)
        return preferences

    def update_preferences(self, preferences: Dict[str, Any]):
        if not self.settings:
            return

        for key, value in preferences.items():
            gsettings_key = key.replace('_', '-')
            if gsettings_key in self.settings.list_keys():
                self._set_setting(gsettings_key, value)

File: src/gear/preferences_manager.rs
use gtk::gio;
use std::collections::HashMap;
use gtk::gio::prelude::*;
// use adw::prelude::*;
use serde::{Deserialize, Serialize};

#[derive(Debug, Clone, Copy, PartialEq, Eq, Serialize, Deserialize)]
pub enum EditorChoice {
    #[serde(rename = "system_default")]
    SystemDefault,
    #[serde(rename = "gedit")]
    Gedit,
    #[serde(rename = "nano")]
    Nano,
    #[serde(rename = "vim")]
    Vim,
    #[serde(rename = "emacs")]
    Emacs,
    #[serde(rename = "code")]
    VsCode,
    #[serde(rename = "atom")]
    Atom,
    #[serde(rename = "subl")]
    Sublime,
    #[serde(rename = "custom")]
    Custom,
}

#[allow(dead_code)]
#[derive(Debug, Clone, Copy, PartialEq, Eq, Serialize, Deserialize)]
pub enum UpstreamFrequency {
    #[serde(rename = "never")]
    Never,
    #[serde(rename = "daily")]
    Daily,
    #[serde(rename = "weekly")]
    Weekly,
    #[serde(rename = "monthly")]
    Monthly,
    #[serde(rename = "on_startup")]
    OnStartup,
    #[serde(rename = "manual")]
    Manual,
}

#[derive(Debug, Clone, Copy, PartialEq, Eq, Serialize, Deserialize)]
pub enum AURConfidenceLevel {
    #[serde(rename = "paranoid")]
    Paranoid,
    #[serde(rename = "conservative")]
    Conservative,
    #[serde(rename = "balanced")]
    Balanced,
    #[serde(rename = "trusting")]
    Trusting,
    #[serde(rename = "permissive")]
    Permissive,
}

#[derive(Debug, Clone, Copy, PartialEq, Eq, Serialize, Deserialize)]
pub enum ThemeMode {
    #[serde(rename = "system")]
    System,
    #[serde(rename = "light")]
    Light,
    #[serde(rename = "dark")]
    Dark,
}

#[allow(dead_code)]
#[derive(Debug, Clone, Copy, PartialEq, Eq, Serialize, Deserialize)]
pub enum UpdateNotificationLevel {
    #[serde(rename = "none")]
    None,
    #[serde(rename = "critical")]
    Critical,
    #[serde(rename = "important")]
    Important,
    #[serde(rename = "all")]
    All,
}

#[derive(Debug, Clone, Copy, PartialEq, Eq, Serialize, Deserialize)]
pub enum LogLevel {
    #[serde(rename = "debug")]
    Debug,
    #[serde(rename = "info")]
    Info,
    #[serde(rename = "warning")]
    Warning,
    #[serde(rename = "error")]
    Error,
}

#[derive(Debug)]
pub struct PreferencesManager {
    settings: Option<gio::Settings>,
}

impl PreferencesManager {
    const SCHEMA_ID: &'static str = "org.gnome.paru-gui";

    pub fn new() -> Self {
        let settings = gio::Settings::new(Self::SCHEMA_ID);
        Self { settings: Some(settings) }
    }

    fn _get_string(&self, key: &str, default: &str) -> String {
        self.settings.as_ref()
            .map(|s| s.string(key).to_string())
            .unwrap_or_else(|| default.to_string())
    }

    fn _set_string(&self, key: &str, value: &str) -> bool {
        self.settings.as_ref()
            .map(|s| s.set_string(key, value).is_ok())
            .unwrap_or(false)
    }

    fn _get_boolean(&self, key: &str, default: bool) -> bool {
        self.settings.as_ref()
            .map(|s| s.boolean(key))
            .unwrap_or(default)
    }

    fn _set_boolean(&self, key: &str, value: bool) -> bool {
        self.settings.as_ref()
            .map(|s| s.set_boolean(key, value).is_ok())
            .unwrap_or(false)
    }

    fn _get_int(&self, key: &str, default: i32) -> i32 {
        self.settings.as_ref()
            .map(|s| s.int(key))
            .unwrap_or(default)
    }

    fn _set_int(&self, key: &str, value: i32) -> bool {
        self.settings.as_ref()
            .map(|s| s.set_int(key, value).is_ok())
            .unwrap_or(false)
    }

    fn _get_strv(&self, key: &str) -> Vec<String> {
        self.settings.as_ref()
            .map(|s| s.strv(key).iter().map(|s| s.to_string()).collect())
            .unwrap_or_default()
    }

    fn _set_strv(&self, key: &str, value: &[&str]) -> bool {
        self.settings.as_ref()
            .map(|s| s.set_strv(key, value).is_ok())
            .unwrap_or(false)
    }

    // Properties
    pub fn first_run(&self) -> bool { self._get_boolean("first-run", true) }
    pub fn set_first_run(&self, value: bool) { self._set_boolean("first-run", value); }

    pub fn simplified_mode(&self) -> bool { self._get_boolean("simplified-mode", true) }
    pub fn set_simplified_mode(&self, value: bool) { self._set_boolean("simplified-mode", value); }

    pub fn default_editor(&self) -> EditorChoice {
        match self._get_string("default-editor", "gedit").as_str() {
            "system_default" => EditorChoice::SystemDefault,
            "gedit" => EditorChoice::Gedit,
            "nano" => EditorChoice::Nano,
            "vim" => EditorChoice::Vim,
            "emacs" => EditorChoice::Emacs,
            "code" => EditorChoice::VsCode,
            "atom" => EditorChoice::Atom,
            "subl" => EditorChoice::Sublime,
            _ => EditorChoice::Custom,
        }
    }
    pub fn set_default_editor(&self, value: EditorChoice) {
        let val = match value {
            EditorChoice::SystemDefault => "system_default",
            EditorChoice::Gedit => "gedit",
            EditorChoice::Nano => "nano",
            EditorChoice::Vim => "vim",
            EditorChoice::Emacs => "emacs",
            EditorChoice::VsCode => "code",
            EditorChoice::Atom => "atom",
            EditorChoice::Sublime => "subl",
            EditorChoice::Custom => "custom",
        };
        self._set_string("default-editor", val);
    }

    pub fn window_width(&self) -> i32 { self._get_int("window-width", 1200) }
    pub fn set_window_width(&self, value: i32) { self._set_int("window-width", value); }

    pub fn window_height(&self) -> i32 { self._get_int("window-height", 800) }
    pub fn set_window_height(&self, value: i32) { self._set_int("window-height", value); }

    pub fn recent_directories(&self) -> Vec<String> { self._get_strv("recent-directories") }
    pub fn set_recent_directories(&self, value: &[&str]) { self._set_strv("recent-directories", value); }

    pub fn max_recent_directories(&self) -> i32 { self._get_int("max-recent-directories", 10) }
    pub fn set_max_recent_directories(&self, value: i32) { self._set_int("max-recent-directories", value); }

    pub fn upstream_check_frequency(&self) -> i32 { self._get_int("upstream-check-frequency", 24) }
    pub fn set_upstream_check_frequency(&self, value: i32) { self._set_int("upstream-check-frequency", value); }

    pub fn upstream_ignore_prerelease(&self) -> bool { self._get_boolean("upstream-ignore-prerelease", true) }
    pub fn set_upstream_ignore_prerelease(&self, value: bool) { self._set_boolean("upstream-ignore-prerelease", value); }

    pub fn upstream_notify_security_only(&self) -> bool { self._get_boolean("upstream-notify-security-only", false) }
    pub fn set_upstream_notify_security_only(&self, value: bool) { self._set_boolean("upstream-notify-security-only", value); }

    pub fn aur_show_trust_icons(&self) -> bool { self._get_boolean("aur-show-trust-icons", true) }
    pub fn set_aur_show_trust_icons(&self, value: bool) { self._set_boolean("aur-show-trust-icons", value); }

    pub fn aur_confidence_level(&self) -> AURConfidenceLevel {
        match self._get_string("aur-confidence-level", "balanced").as_str() {
            "paranoid" => AURConfidenceLevel::Paranoid,
            "conservative" => AURConfidenceLevel::Conservative,
            "balanced" => AURConfidenceLevel::Balanced,
            "trusting" => AURConfidenceLevel::Trusting,
            "permissive" => AURConfidenceLevel::Permissive,
            _ => AURConfidenceLevel::Balanced,
        }
    }
    pub fn set_aur_confidence_level(&self, value: AURConfidenceLevel) {
        let val = match value {
            AURConfidenceLevel::Paranoid => "paranoid",
            AURConfidenceLevel::Conservative => "conservative",
            AURConfidenceLevel::Balanced => "balanced",
            AURConfidenceLevel::Trusting => "trusting",
            AURConfidenceLevel::Permissive => "permissive",
        };
        self._set_string("aur-confidence-level", val);
    }

    pub fn developer_mode(&self) -> bool { self._get_boolean("developer-mode", false) }
    pub fn set_developer_mode(&self, value: bool) { self._set_boolean("developer-mode", value); }

    pub fn clean_after_build(&self) -> bool { self._get_boolean("clean-after-build", true) }
    pub fn set_clean_after_build(&self, value: bool) { self._set_boolean("clean-after-build", value); }

    pub fn enable_sandboxing(&self) -> bool { self._get_boolean("enable-sandboxing", true) }
    pub fn set_enable_sandboxing(&self, value: bool) { self._set_boolean("enable-sandboxing", value); }

    pub fn theme_mode(&self) -> ThemeMode {
        match self._get_string("theme-mode", "system").as_str() {
            "light" => ThemeMode::Light,
            "dark" => ThemeMode::Dark,
            _ => ThemeMode::System,
        }
    }
    pub fn set_theme_mode(&self, value: ThemeMode) {
        let val = match value {
            ThemeMode::System => "system",
            ThemeMode::Light => "light",
            ThemeMode::Dark => "dark",
        };
        self._set_string("theme-mode", val);
    }

    pub fn log_level(&self) -> LogLevel {
        match self._get_string("log-level", "info").as_str() {
            "debug" => LogLevel::Debug,
            "info" => LogLevel::Info,
            "warning" => LogLevel::Warning,
            "error" => LogLevel::Error,
            _ => LogLevel::Info,
        }
    }
    pub fn set_log_level(&self, value: LogLevel) {
        let val = match value {
            LogLevel::Debug => "debug",
            LogLevel::Info => "info",
            LogLevel::Warning => "warning",
            LogLevel::Error => "error",
        };
        self._set_string("log-level", val);
    }

    pub fn add_recent_directory(&self, directory: &str) {
        let mut recent = self.recent_directories();
        if let Some(pos) = recent.iter().position(|d| d == directory) {
            recent.remove(pos);
        }
        recent.insert(0, directory.to_string());
        let max = self.max_recent_directories() as usize;
        if recent.len() > max {
            recent.truncate(max);
        }
        let refs: Vec<&str> = recent.iter().map(|s| s.as_str()).collect();
        self.set_recent_directories(&refs);
    }

    pub fn clear_recent_directories(&self) {
        self.set_recent_directories(&[]);
    }

    pub fn get_editor_command(&self) -> String {
        let editor = self.default_editor();
        match editor {
            EditorChoice::SystemDefault => std::env::var("EDITOR").unwrap_or_else(|_| "gedit".to_string()),
            EditorChoice::Custom => self._get_string("custom-editor-command", "gedit"),
            EditorChoice::Gedit => "gedit".to_string(),
            EditorChoice::Nano => "nano".to_string(),
            EditorChoice::Vim => "vim".to_string(),
            EditorChoice::Emacs => "emacs".to_string(),
            EditorChoice::VsCode => "code".to_string(),
            EditorChoice::Atom => "atom".to_string(),
            EditorChoice::Sublime => "subl".to_string(),
        }
    }

    pub fn reset_to_defaults(&self) -> bool {
        /*
        if let Some(settings) = &self.settings {
            for key in settings.keys() {
                settings.reset(&key);
            }
            return true;
        }
        */
        false
    }

    pub fn validate_preferences(&self) -> HashMap<String, Vec<String>> {
        let mut errors = HashMap::new();

        if self.window_width() < 800 {
            errors.entry("window".to_string()).or_insert_with(Vec::new).push("Window width must be at least 800 pixels".to_string());
        }

        if self.window_height() < 600 {
            errors.entry("window".to_string()).or_insert_with(Vec::new).push("Window height must be at least 600 pixels".to_string());
        }

        if self.upstream_check_frequency() < 0 {
            errors.entry("upstream".to_string()).or_insert_with(Vec::new).push("Check frequency cannot be negative".to_string());
        }

        errors
    }

    pub fn apply_theme(&self) {
        let theme = self.theme_mode();
        if theme == ThemeMode::System { return; }

        let style_manager = adw::StyleManager::default();
        match theme {
            ThemeMode::Dark => style_manager.set_color_scheme(adw::ColorScheme::ForceDark),
            ThemeMode::Light => style_manager.set_color_scheme(adw::ColorScheme::ForceLight),
            _ => style_manager.set_color_scheme(adw::ColorScheme::Default),
        }
    }
}
